feat: Upgrade support for Julia

savq opened the pull request:
This PR upgrades support for Julia to a more recent version of tree-sitter-julia.

- See tree-sitter/tree-sitter-julia#135 for the changes to the parser
- See semgrep/ocaml-tree-sitter-semgrep#520 for the changes to semgrep-julia

Some notes:

- The changes to tree-sitter-julia in the linked PR were focused on producing parse trees more faithful to JuliaSyntax (the default Julia parser). However, JuliaSyntax delegates a lot of work to the lowering phase, like checking function signatures. This means that while tree-sitter produces a smaller parser, semgrep now needs to do more work to find a function's name and parameters.

- Since parameters are now parsed as expressions, I left a lot of `todo` for expressions that don't constitute valid "patterns". I don't know if it might be a good idea to create an alias for `todo` to clearly indicate that we don't handle those cases because they're not valid in normal Julia (outside macros), and not because they should be handled in the future.

- tree-sitter-julia now parses short functions as assignments (That's also how it's done in the reference parser). To handle this, I translate short functions into lambdas: `f(...) = ...` becomes `f = (...) -> ...`.

- The special handling for `where_clause` was removed. A later version of tree-sitter-julia unified `where_clause` with `where_expression`.

- One of the issues I really hoped would be fixed was #10649, but that's actually a limitation on semgrep's side! In general, I still feel there's a lot of weird hacks because semgrep's AST only considers C-like macros.

savq committed:
Rename bare_tuple to open_tuple

savq committed:
Remove _number

savq committed:
Remove _quotable

savq committed:
Refactor _operation

savq committed:
Refactor abstract/primitive type tokens

savq committed:
Refactor function signatures

Try to do as much as possible in `map_signature` to reduce duplication
between function, macro and "short function" definitions.

savq committed:
Refactor parameter and argument lists

savq committed:
Refactor assignments and short functions

tree-sitter-julia now parses short functions as assignments
(That's also how it's done in the reference parser).
To handle this, transform short function into lambda:
`f(...) = ...` becomes `f = (...) -> ...`.

savq committed:
Refactor parameters

Julia doesn't distinguish expressions and patterns. To parse parameters
we must pretend that some expressions are patterns where only a few
syntactic forms are valid.

savq committed:
Fix bracket level expressions and other stuff

- Fix some issues with bracket level expressions, let bindings and named
  fields.
- Fix do parameters
- Fix macro parameters
- Fix the rename of `relative_qualifier` to `Import_path`.

savq committed:
Remove unused functions

Currently we don't handle `where_clause`. That's mostly fine since
later versions of tree-sitter-julia don't have that rule
(`where_expression` is used instead).

savq committed:
Fix open macrocall expression as parameter

nmote commented:
Do I understand correctly that this updates Semgrep to pull in the changes in https://github.com/semgrep/ocaml-tree-sitter-semgrep/pull/520, but without actually updating the submodule?

I'm assuming yes, so I'll review that PR and release `semgrep-julia` with the changes assuming it all looks good.

aviks commented:
Hi @nmote do you need anything else before being able to review this change? Thanks!

nmote commented:
@aviks Yes, an answer to my question in this thread would be helpful. Because this doesn't actually include a submodule update (understandable since we haven't released semgrep-julia based on the new changes yet), it's hard to tell for sure exactly which changes this pulls in, and we can't yet run CI to test.

I do intend to review the ocaml-tree-sitter-semgrep PR either way, but I've been busy and the lack of a response here led me to prioritize it below other urgent work items.

savq commented:
Hi @nmote, from what I understood when I asked about this on the semgrep slack, the PR on semgrep/ocaml-tree-sitter-semgrep should be merged before the PR here. Should I update the submodule to point to my working branch?

nmote commented:
Nope it's fine, I just wanted to confirm that I understood the situation correctly! Have I?

savq commented:
Yes, I worked on these changes locally with the submodule pointing to the updated branch from the other repo. Like you said, it'd be impossible to test otherwise. 😅

nmote commented:
Sounds good, thanks for clarifying! I'll review both PRs today. Apologies for the delay.

nmote commented:
This looks good, thanks for the contribution! Assuming tests pass once you pull in `semgrep-swift`, I am mostly ready to merge it, but I am a bit concerned by the number of calls to `todo` throughout. I'm not sure whether you saw, but that simply calls `failwith` and if we encounter any of those, we will abort parsing the entire file. It's best to avoid that. Even if they are "impossible" cases, future changes may make them more possible, and I've also seen tree sitter error recovery generate CSTs that we thought impossible.

Given that Julia is an experimental language for Semgrep, I could be convinced to merge this with those in place as long as tests pass, but I'd much prefer something safer. Could you do a pass through to determine how many of those you can reasonably replace with a real AST node? Even the `Other` variants are fine for rare-ish cases. You could also try using the raw tree type. Let me know whether that sounds feasible.

savq commented:
Some todos represent impossible cases: invalid Julia syntax that tree-sitter allows because disallowing them is a bigger hurdle (usually a big increase in the number of parse states).

Other todos are technically valid syntax, but in practice they're only usable in macros. I'll add more comments to clarify, and check which of these can be turned to `Other` variants.

nmote commented:
Sounds good! Let me know when you are done and I will take another look.

nmote reviewed (commented):
Adding a review to get this out of my queue. Please re-request review from me once you're ready or if you have questions or comments that you'd like to make sure don't fall through the cracks. Alternatively, feel free to ping me on community slack.

savq committed:
Add comments explaining todo calls

Comments with `NOTE(syntax)` explain things that happen at the parser
level. If some form is completely invalid, it's a syntax problem
(tree-sitter-julia might allow these for the sake of simplicity).

Comments with `NOTE(lowering)` explain things that happen at the
lowering level/after parsing. If some form can be quoted, but not
evaluated, it's a lowering problem (tree-sitter-julia allows these for
macros).

savq commented:
I added comments explaining the relevant `todo` calls. I only found one `todo` (related to type parameters) that was important to fix.

nmote commented:
Thanks! I also need you to pull in the `semgrep-julia` release I made based on your other PR, so that we can run the tests. I'll review today but I do need to see CI green before merging.

nmote reviewed (commented):
Thanks! I'm still not entirely comfortable with the number of `todo`s for the reasons I outlined above, but given that Julia is experimental I am willing to merge this nonetheless. If we start getting bug reports related to these `todo`s I'll have to revert, though.

A couple last items:
* Please pull in `semgrep-swift` so that we can run the tests.
* Please add a changelog entry (see the `changelog.d` file). Something like "Upgraded the Julia parser to the latest version of tree-sitter-julia" would work.

Re-request once you've done this and I'll give it a (hopefully) last look!

Thanks again for contributing.

savq committed:
Update semgrep-julia submodule

savq committed:
Add changelog entry for julia upgrade

savq commented:
> I'm still not entirely comfortable with the number of todos 

Yeah, I understand. There are definitely things to improve. I made a [PR to tree-sitter-julia](https://github.com/tree-sitter/tree-sitter-julia/pull/170) to address some of the difficulties I had while working on this ("anonymous choices", some overly specific rules) so the next PR should simplify things further and remove the need for the todos in the first place. 

Thanks for the review!

nmote commented:
Hey, my apologies: there's also a Swift upgrade in progress and I mistyped. I meant to ask you to update `semgrep-julia`. In fact, this PR should not include a `semgrep-swift` submodule update.

This looks good, and I'm ready to merge. But could you undo the `semgrep-swift` update? Once you do and we see a good CI run this is ready! Thanks again!

savq commented:
> this PR should not include a semgrep-swift submodule update.

Yeah, that makes more sense. I removed the that. 👍

nmote commented:
Sadly, there are quite a few problems visible now that we have CI running. Could you take a look? I'm running a release today but I can answer questions later if you would like a hand.

nmote reviewed (commented):
I'm commenting to get this out of my queue for now. Re-request review when you'd like me to take another look (including answering questions). Also feel free to ping me on the community slack.

savq committed:
Use List_.map instead of List.map

savq committed:
Update semgrep-julia and fix statement ellipsis

semgrep-julia was extending `_statement` with `semgrep_ellipsis`, but
this was redundant since both statements and ellipses are expressions.

savq committed:
Fix signature parsing issue

- Fix regression with metavariables in function definitions.
- Fix semgrep ellipsis in function parameters.

savq committed:
Distinguish top_level in blocks vs elsewhere

Parsing all top_level as semgrep expressions was causing all variable
definitions to be parsed as `Assign` instead of `VarDef`.

nmote committed:
Merge branch 'develop' into upgrade-julia

nmote reviewed (approved):
Looks good, and we had a green CI run, aside from the submodule job that doesn't like forks. We've had a few changes to ocaml-tree-sitter-semgrep that affect code generation (but shouldn't affect the changes here), so I'm going to merge `develop` into your PR there, regenerate `semgrep-julia`, and then update it here. Assuming CI is still green, which I expect it to be, I'll then merge your PRs.

Thanks for contributing!

nmote committed:
Latest semgrep-julia

nmote committed:
Merge branch 'develop' into upgrade-julia
